src/Kubernetes.Controller/Certificates/ServerCertificateSelector.cs

```diff
@@ -1,27 +1,118 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Formats.Asn1;
+using System.Linq;
 using System.Security.Cryptography.X509Certificates;
+using System.Threading;
 using Microsoft.AspNetCore.Connections;
 
 namespace Yarp.Kubernetes.Controller.Certificates;
 
 internal class ServerCertificateSelector : IServerCertificateSelector
 {
-    private X509Certificate2 _defaultCertificate;
+    private readonly Dictionary<NamespacedName, CertificateCandidate> _certificates = new();
+    private readonly ReaderWriterLock _readerWriterLock = new();
 
     public void AddCertificate(NamespacedName certificateName, X509Certificate2 certificate)
     {
-        _defaultCertificate = certificate;
+        _readerWriterLock.AcquireWriterLock(Timeout.Infinite);
+        try
+        {
+            _certificates[certificateName] = new(certificate);
+        }
+        finally
+        {
+            _readerWriterLock.ReleaseWriterLock();
+        }
     }
 
     public X509Certificate2 GetCertificate(ConnectionContext connectionContext, string domainName)
     {
-        return _defaultCertificate;
+        _readerWriterLock.AcquireReaderLock(Timeout.Infinite);
+        try
+        {
+            foreach (var (cert, domains) in _certificates.Values)
+            {
+                if (domains.Contains(domainName))
+                {
+                    return cert;
+                }
+            }
+        }
+        finally
+        {
+            _readerWriterLock.ReleaseReaderLock();
+        }
+
+        return _certificates.Values.FirstOrDefault()?.Certificate;
     }
 
     public void RemoveCertificate(NamespacedName certificateName)
     {
-        _defaultCertificate = null;
+        _readerWriterLock.AcquireWriterLock(Timeout.Infinite);
+        try
+        {
+            _certificates.Remove(certificateName);
+        }
+        finally
+        {
+            _readerWriterLock.ReleaseWriterLock();
+        }
+    }
+
+    private record CertificateCandidate
+    {
+        public IReadOnlySet<string> Domains { get; }
+
+        public X509Certificate2 Certificate { get; }
+
+        public CertificateCandidate(X509Certificate2 certificate)
+        {
+            Certificate = certificate;
+            Domains = ImmutableHashSet.CreateRange(StringComparer.InvariantCultureIgnoreCase, GetDomains(certificate));
```

Comment:
```suggestion
            Domains = new HashSet(GetDomains(certificate), StringComparer.OrdinalIgnoreCase);
```
Invariant culture is rarely the right choice. `ImmutableHashSet` is also unnecessary overhead in this case since this is all internal logic.